packages/rich-text/src/test/insert-object.js

```diff
@@ -17,12 +17,16 @@ describe( 'insert', () => {
 	it( 'should delete and insert', () => {
 		const record = {
 			formats: [ , , , , [ em ], [ em ], [ em ], , , , , , , ],
+			lines: [ , , , , , , , , , , , , , ],
+			objects: [ , , , , , , , , , , , , , ],
 			text: 'one two three',
 			start: 6,
 			end: 6,
 		};
 		const expected = {
-			formats: [ , , [ { ...obj, object: true } ], [ em ], , , , , , , ],
+			formats: [ , , , [ em ], , , , , , , ],
```

Comment:
I don't know if / how a browser optimizes memory management for allocated arrays, but it appears wasteful to me to create three distinct arrays for each paragraph with a size corresponding to the number of characters (?) in the paragraph. Particularly when it seems that for a given "index", there could be only one qualifying option amongst (a) an array of formats, (b) an object definition, (c) a line descriptor (?). Would it not be possible to represent this as a single array? I guess I'd see it as much the same as it was before, just not called `formats`, and where each "sort" is treated differently. Or is that the root of the issue being addressed?

Reply:
See also https://github.com/WordPress/gutenberg/pull/13948#issuecomment-468957243.

I think it makes sense regardless of memory use to keep it at one sparse array.

I've also dug a bit further into the memory usage of sparse arrays. Some resources suggest that the browser optimises for these. I checked in Chrome, and it appears that each empty slot takes up 8 bytes.

```
Array( 0 ) = 208 bytes
Array( 1 ) = 232 bytes
Array( 2 ) = 240 bytes
Array( 3 ) = 248 bytes
```

This is very different from how I thought about it. I thought the array length was stored as a number, and empty slots would not take up any space. If it's true that each empty slot takes up 8 bytes per character, then an empty sparse array takes 4 times the space of the text string.

Reply:
Of course, if we merge the three sparse arrays into one, we'd also create extra objects or arrays, so from a certain amount of formats, the size will be bigger.